stream-parent/stream-receiver/src/main/java/org/interledger/stream/server/SpspStreamConnectionGenerator.java

```diff
@@ -0,0 +1,112 @@
+  private static final byte[] STREAM_SERVER_SECRET_GENERATOR = "ilp_stream_secret_generator".getBytes(US_ASCII);
+
+  @Override
+  public StreamConnectionDetails generateConnectionDetails(
+      final ServerSecretSupplier serverSecretSupplier, final InterledgerAddress receiverAddress
+  ) {
+    Objects.requireNonNull(serverSecretSupplier, "serverSecretSupplier must not be null");
+    Objects.requireNonNull(receiverAddress, "receiverAddress must not be null");
+    Preconditions.checkArgument(serverSecretSupplier.get().length >= 32, "Server secret must be 32 bytes");
+
+    // base_address + "." + 32-bytes encoded as base64url
+    final Builder streamConnectionDetailsBuilder = StreamConnectionDetails.builder();
+
+    final byte[] randomBytes = Random.randBytes(18);
+    final byte[] sharedSecret = Hashing
+        .hmacSha256(secretGenerator(serverSecretSupplier))
+        .hashBytes(randomBytes)
+        .asBytes();
+
+    final String destinationAddressPrecursor =
+        receiverAddress.with(Base64.getUrlEncoder().withoutPadding().encodeToString(randomBytes)).getValue();
+
+    // The authTag is the first 14 bytes of the HmacSha256 of destinationAddressPrecursor
+    final byte[] authTag = Arrays.copyOf(
+        Hashing.hmacSha256(sharedSecret)
+            .hashBytes(destinationAddressPrecursor.getBytes(US_ASCII))
+            .asBytes(),
+        14
+    );
+
+    final InterledgerAddress destinationAddress = InterledgerAddress.of(
+        destinationAddressPrecursor + Base64.getUrlEncoder().withoutPadding().encodeToString(authTag)
+    );
+
+    return streamConnectionDetailsBuilder
+        .sharedSecret(Base64.getUrlEncoder().withoutPadding().encodeToString(sharedSecret))
+        .destinationAddress(destinationAddress)
+        .build();
+  }
+
+  @Override
+  public byte[] deriveSecretFromAddress(
+      final ServerSecretSupplier serverSecretSupplier, final InterledgerAddress receiverAddress
+  ) {
+    Objects.requireNonNull(receiverAddress);
+
+    final String receiverAddressAsString = receiverAddress.getValue();
+    final String localPart = receiverAddressAsString.substring(receiverAddressAsString.lastIndexOf(".") + 1);
+
+    final byte[] localPartBytes = Base64.getUrlDecoder().decode(localPart);
+    if (localPartBytes.length != 32) {
+      throw new StreamException(
+          String.format("Invalid Receiver Address (should have been 32 byte long): %s", receiverAddress));
+    }
+
+    // Bytes 0 through 17
+    final byte[] randomBytes = Arrays.copyOf(localPartBytes, 18);
+    final byte[] sharedSecret = Hashing.hmacSha256(secretGenerator(serverSecretSupplier)).hashBytes(randomBytes)
+        .asBytes();
+    // Bytes 18 through 31
+    final byte[] authTag = Arrays.copyOfRange(localPartBytes, 18, localPartBytes.length);
+
+    // The Address without the final 18 bytes.
+    String addressWithoutBytes = receiverAddressAsString.substring(0, receiverAddressAsString.length() - 19);
```

Comment:
Do all these `final` really add much :) ? 
I have been programming in TypeScript/JavaScript mostly lately, and yeah, I habitually use const where possible. The alternative is `var` or `let`, where you have to use *something* to declare a reference, whereas with Java you *can* eschew it.  

Does Java not have a tool to add them in automatically where they are "effectively final" ? Seems like a pain. 

Context being that seems `addressWithoutBytes` seems to be lacking `final`.

Reply:
> Do all these final really add much :) ?

Not really. It's _mostly_ a style thing. E.g., here's a [pretty informative SO](https://stackoverflow.com/questions/8354412/do-java-finals-help-the-compiler-create-more-efficient-bytecode) about why `final` is mostly not useful, but not harmful. 

I personally like `final` modifiers, but at this point in my career I consider it a benign habit that's just part of my style. Mostly I like them because it helps me not have to think about potential simple bugs, like  accidentally mutating a reference somewhere "below" its declaration.

Also, I like them as modifiers to methods, especially constructors, in order to avoid mistakes in assignment. E.g., if you forget to use `this` in the constructor assignment example below, this is a silly bug that the compiler will catch:

```
class Foo {
  private bar;
  Foo(final int bar){
    bar = bar;
  }
}
```

However, the compiler won't catch this mistake if the code is structured like this:

```
class Foo {
  private bar;
  Foo(int bar){
    bar = bar;
  }
}
```

There of plenty of people I highly respect, however, that will offer counter-arguments like, "but modern IDEs will warn on this condition" or "the JIT will actually ignore your modifiers if it makes the code more optimal" or "This makes the code more difficult to read" or whatever.

Like I said, this is purely a style thing, so using `final` or not using it makes _almost_ no difference in practice, especially if code is well tested.